### PR TITLE
Fix FormField implementation components TextFormField, DropdownButtonFormField, CupertinoTextFormFieldRow. The problem that their onChange() method cannot be executed when the parent Form component calls the reset() method

### LINKED ISSUE #123009: When the Form widget calls the reset method, if it contains widget of the FormField type (such as TextFormField), the onChange method of those components will not be called
```
class TestWidget extends StatelessWidget {
  final _formKey = GlobalKey<FormState>();
  TestWidget({Key? key}) : super(key: key);
  
  @override
  Widget build(BuildContext context) {
    return Form(
      key: _formKey,
      child: Column(
        children: [
          ElevatedButton(
            onPressed: (){
              _formKey.currentState?.reset();
            },
            child: const Text("reset")
          ),
          TextFormField(
            onChanged: (value) {
              // not executed on _formKey.currentState?.reset()
              print(value);
            }
          )
        ],
      )
    );
  }
}
```

### PR DESCRIPTION
**Pull request by yeleibo**
fixes https://github.com/flutter/flutter/issues/123009
<details>
<summary>code sample</summary>
```

    import 'package:flutter/cupertino.dart';
    import 'package:flutter/material.dart';
    
    void main() => runApp(const MyApp());
    
    class MyApp extends StatelessWidget {
      const MyApp({super.key});
    
      @override
      Widget build(BuildContext context) {
        return MaterialApp(
          debugShowCheckedModeBanner: false,
          theme: ThemeData(useMaterial3: true),
          home: Example(),
        );
      }
    }
    
    class Example extends StatelessWidget {
      final _formKey = GlobalKey<FormState>();
      Example({super.key});
    
      @override
      Widget build(BuildContext context) {
        return Scaffold(
          body: Form(
              key: _formKey,
              child: Column(
                children: [
                  CupertinoTextFormFieldRow(onChanged: (value) {
                    print("CupertinoTextFormFieldRow widget onChange execute");
                  }),
                  TextFormField(onChanged: (value) {
                    print("TextFormField widget onChange execute");
                  }),
                  DropdownButtonFormField(
                    items: const [
                      DropdownMenuItem(
                        value: "item1",
                        child: Text("item1"),
                      )
                    ],
                    onChanged: (value) {
                      print("DropdownButtonFormField widget onChange execute");
                    },
                  ),
                  ElevatedButton(
                      onPressed: () {
                        _formKey.currentState?.reset();
                      },
                      child: const Text("reset")),
                ],
              )),
        );
      }
    }

</details>

### CODE REVIEW

File: packages/flutter/lib/src/cupertino/text_form_field_row.dart [hunk 1]
```diff
@@ -334,6 +335,9 @@ class _CupertinoTextFormFieldRowState extends FormFieldState<String> {
         _effectiveController!.text = widget.initialValue!;
       });
     }
+    _cupertinoTextFormFieldRow.onChanged?.call(_effectiveController!.text);
+
+
```
Comment (justinmc) [inline packages/flutter/lib/src/cupertino/text_form_field_row.dart:340]: Nit: Too many newlines here.

File: packages/flutter/lib/src/material/dropdown.dart [hunk 2]
```diff
@@ -1690,4 +1690,11 @@ class _DropdownButtonFormFieldState<T> extends FormFieldState<T> {
       setValue(widget.initialValue);
     }
   }
+  @override
```
Comment (justinmc) [inline packages/flutter/lib/src/material/dropdown.dart:1693]: Nit: Put a newline above the annotation.

**Review by justinmc — COMMENTED** [review]
Thanks for the PR!  Are you able to add a test for this?

**Comment by yeleibo** [comment]
> Thanks for the PR! Are you able to add a test for this?
Yes, tests have been added. 
The Linux docs_test action check is not passing, but I'm not sure why.

File: packages/flutter/lib/src/material/text_form_field.dart [hunk 3]
```diff
@@ -248,7 +248,8 @@ class TextFormField extends FormField<String> {
   /// If null, this widget will create its own [TextEditingController] and
   /// initialize its [TextEditingController.text] with [initialValue].
   final TextEditingController? controller;
-
+  /// {@macro flutter.material.TextField.onChanged}
```
Comment (justinmc) [inline packages/flutter/lib/src/material/text_form_field.dart:251]: This macro doesn't exist, did you mean `flutter.material.editableText.onChanged`?  Or you could create this macro if you want the entire TextField.onChanged docs.

File: packages/flutter/lib/src/cupertino/text_form_field_row.dart [hunk 4]
```diff
@@ -259,7 +259,8 @@ class CupertinoTextFormFieldRow extends FormField<String> {
   /// If null, this widget will create its own [TextEditingController] and
   /// initialize its [TextEditingController.text] with [initialValue].
   final TextEditingController? controller;
-
+  /// {@macro flutter.material.CupertinoFormRow.onChanged}
```
Comment (justinmc) [inline packages/flutter/lib/src/cupertino/text_form_field_row.dart:262]: This macro also doesn't seem to exist.

File: packages/flutter/test/cupertino/text_form_field_row_test.dart [hunk 5]
```diff
@@ -492,4 +492,39 @@ void main() {
     final CupertinoTextField rtlTextFieldWidget = tester.widget(rtlTextFieldFinder);
     expect(rtlTextFieldWidget.textDirection, TextDirection.rtl);
   });
+
+  testWidgets('CupertinoTextFormFieldRow onChanged to initialValue when Form reset', (WidgetTester tester) async {
+    final GlobalKey<FormFieldState<String>> stateKey = GlobalKey<FormFieldState<String>>();
+    final GlobalKey<FormState> formKey = GlobalKey<FormState>();
+    String value='initialValue';
```
Comment (justinmc) [inline packages/flutter/test/cupertino/text_form_field_row_test.dart:499]:
```suggestion
    String value = 'initialValue';
```

Comment (justinmc) [reply packages/flutter/test/cupertino/text_form_field_row_test.dart:499]: And anywhere else you're not spacing this out needs to have spaces too.

File: packages/flutter/test/cupertino/text_form_field_row_test.dart [hunk 6]
```diff
@@ -492,4 +492,39 @@ void main() {
...
+              },
+            ),
+          ),
+        ),
+      ),
+    );
+
+    expect(stateKey.currentState!.value, 'initialValue');
+    expect(value, 'initialValue');
+    //change value to 'changedValue'
```
Comment (justinmc) [inline packages/flutter/test/cupertino/text_form_field_row_test.dart:520]:
```suggestion
    // Change value to 'changedValue'.
```

And elsewhere you have comments without punctuation and capital letters.

**Review by justinmc — COMMENTED** [review]
Thanks for adding the tests!  I think the docs failures have to do with macros that don't exist.  I commented about that below.

**Comment by justinmc** [comment]
Looks like there are still docs test failures, it mentions `[onSubmitted]`.

File: packages/flutter/lib/src/cupertino/text_form_field_row.dart [hunk 7]
```diff
@@ -260,6 +260,9 @@ class CupertinoTextFormFieldRow extends FormField<String> {
   /// initialize its [TextEditingController.text] with [initialValue].
   final TextEditingController? controller;
 
+  /// Called when the user initiates a change to the TextField's
+  /// value: when they have inserted or deleted text or form reset.
+  final ValueChanged<String>? onChanged;
```
Comment (justinmc) [inline packages/flutter/lib/src/cupertino/text_form_field_row.dart:265]: Nit: Add a newline after this.

File: packages/flutter/lib/src/material/text_form_field.dart [hunk 8]
```diff
@@ -249,6 +249,9 @@ class TextFormField extends FormField<String> {
   /// initialize its [TextEditingController.text] with [initialValue].
   final TextEditingController? controller;
 
+  /// Called when the user initiates a change to the TextField's
+  /// value: when they have inserted or deleted text or form reset.
+  final ValueChanged<String>? onChanged;
```
Comment (justinmc) [inline packages/flutter/lib/src/material/text_form_field.dart:254]: Nit: Newline here too.

File: packages/flutter/test/cupertino/text_form_field_row_test.dart [hunk 9]
```diff
@@ -492,4 +492,39 @@ void main() {
...
+      ),
+    );
+
+    expect(stateKey.currentState!.value, 'initialValue');
+    expect(value, 'initialValue');
+    // Change value to 'changedValue'.
+    await tester.enterText(find.byType(CupertinoTextField), 'changedValue');
+    expect(stateKey.currentState!.value,'changedValue');
+    expect(value, 'changedValue');
+    // form to reset
```
Comment (justinmc) [inline packages/flutter/test/cupertino/text_form_field_row_test.dart:524]: Change to: "Reset form."

File: packages/flutter/test/material/dropdown_test.dart [hunk 10]
```diff
@@ -3955,4 +3955,48 @@ void main() {
...
+              },
+            ),
+          ),
+        ),
+      ),
+    );
+
+    expect(value, isNull);
+    expect(stateKey.currentState!.value, equals('One'));
+    //value chang to Two
```
Comment (justinmc) [inline packages/flutter/test/material/dropdown_test.dart:3990]: `// Change value to "Two".`

File: packages/flutter/test/material/dropdown_test.dart [hunk 11]
```diff
@@ -3955,4 +3955,48 @@ void main() {
...
+    await tester.tap(find.text('One'));
+    await tester.pumpAndSettle();
+    await tester.tap(find.text('Two').last);
+    await tester.pumpAndSettle();
+    expect(value, equals('Two'));
+    expect(stateKey.currentState!.value, equals('Two'));
+    // form to reset
```
Comment (justinmc) [inline packages/flutter/test/material/dropdown_test.dart:3997]: `// Reset form.`

File: packages/flutter/test/material/text_form_field_test.dart [hunk 12]
```diff
@@ -1219,4 +1219,37 @@ void main() {
...
+      ),
+    ));
+
+    expect(stateKey.currentState!.value, 'initialValue');
+    expect(value, 'initialValue');
+    //change value to 'changedValue'
+    await tester.enterText(find.byType(TextField), 'changedValue');
+    expect(stateKey.currentState!.value,'changedValue');
+    expect(value, 'changedValue');
+    // form to reset
```
Comment (justinmc) [inline packages/flutter/test/material/text_form_field_test.dart:1249]: "Reset form."

File: packages/flutter/test/material/text_form_field_test.dart [hunk 13]
```diff
@@ -1219,4 +1219,37 @@ void main() {
...
+              value = newValue;
+            },
+          ),
+        ),
+      ),
+    ));
+
+    expect(stateKey.currentState!.value, 'initialValue');
+    expect(value, 'initialValue');
+    //change value to 'changedValue'
```
Comment (justinmc) [inline packages/flutter/test/material/text_form_field_test.dart:1245]: `// Change value to "changedValue".`

File: packages/flutter/lib/src/cupertino/text_form_field_row.dart [hunk 14]
```diff
@@ -334,6 +337,7 @@ class _CupertinoTextFormFieldRowState extends FormFieldState<String> {
         _effectiveController!.text = widget.initialValue!;
       });
     }
+    _cupertinoTextFormFieldRow.onChanged?.call(_effectiveController!.text);
```
Comment (justinmc) [inline packages/flutter/lib/src/cupertino/text_form_field_row.dart:340]: What if the value didn't actually change, should onChanged still be called?

File: packages/flutter/lib/src/material/dropdown.dart [hunk 15]
```diff
@@ -1685,4 +1685,12 @@ class _DropdownButtonFormFieldState<T> extends FormFieldState<T> {
       setValue(widget.initialValue);
     }
   }
+
+  @override
+  void reset() {
+    final DropdownButtonFormField<T> dropdownButtonFormField = widget as DropdownButtonFormField<T>;
+    assert(dropdownButtonFormField.onChanged != null);
+    dropdownButtonFormField.onChanged!(widget.initialValue);
+    super.reset();
```
Comment (justinmc) [inline packages/flutter/lib/src/material/dropdown.dart:1694]: In CupertinoTextFormFieldRow, `reset` calls `super` on the first line.  Should this one also call it on the first line?

File: packages/flutter/lib/src/material/text_form_field.dart [hunk 16]
```diff
@@ -345,6 +348,7 @@ class _TextFormFieldState extends FormFieldState<String> {
     // setState will be called in the superclass, so even though state is being
     // manipulated, no setState call is needed here.
     _effectiveController.text = widget.initialValue ?? '';
+    _textFormField.onChanged?.call(_effectiveController.text);
     super.reset();
```
Comment (justinmc) [inline packages/flutter/lib/src/material/text_form_field.dart:352]: And this one calls `super` on the last line again.  Not sure which is right 🤷 .

**Review by justinmc — COMMENTED** [review]
Mostly nits here.  Thanks for getting all of the checks to pass.

**Review by justinmc — COMMENTED** [review]
Mostly nits here.  Thanks for getting all of the checks to pass.

**Comment by Hixie** [comment]
@yeleibo Thanks for the contribution! Will you be able to continue working on this? It's totally fine if not; we can take it over (though it might take longer). I'm just asking because I'm trying to clean out our review queue.

**Comment by bleroux** [comment]
Closing this PR because I filed https://github.com/flutter/flutter/pull/134295 based on the work done here.
Thanks @yeleibo for initiating this work!

**Pull request closed without merging**